File: lib/src/main/java/com/king/frame/mvvmframe/base/BaseApplication.java
package com.king.frame.mvvmframe.base;

import android.app.Application;
import android.content.Context;

import com.king.frame.mvvmframe.base.delegate.ApplicationDelegate;
import com.king.frame.mvvmframe.di.component.AppComponent;
import com.king.frame.mvvmframe.util.Preconditions;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;

/**
 *  MVVMFrame 框架基于Google官方的Architecture Components dependencies 构建，在使用MVVMFrame时，需遵循一些规范：
 *  1.你的项目中的Application中需初始化MVVMFrame框架相关信息，有两种方式处理：
 *      a.直接继承本类{@link BaseApplication}即可；
 *      b.如你的项目中的Application本身继承了其它第三方的Application，因为Java是单继承原因，导致没法继承本类，可参照本类，
 *      将{@link #mApplicationDelegate}相关代码复制到你项目的Application中，在相应的生命周期中调用即可。
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public class BaseApplication extends Application implements IAppComponent, HasAndroidInjector {

    /**
     * Application 代理 规避项目中集成了其它第三方类或其它原因，不能集成本类{@link BaseApplication}时，
     * 参照本类实现 {@link ApplicationDelegate} 即可初始化MVVMFrame框架配置信息。
     */
    private ApplicationDelegate mApplicationDelegate;

    /**
     * Dagger.Android 注入
     */
    @Inject
    DispatchingAndroidInjector<Object> mAndroidInjector;


    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        mApplicationDelegate = new ApplicationDelegate(this);
        mApplicationDelegate.attachBaseContext(base);

    }


    @Override
    public void onCreate() {
        super.onCreate();
        mApplicationDelegate.onCreate();

    }

    @Override
    public void onTerminate() {
        super.onTerminate();
        mApplicationDelegate.onTerminate();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mApplicationDelegate.onLowMemory();
    }

    @Override
    public void onTrimMemory(int level) {
        super.onTrimMemory(level);
        mApplicationDelegate.onTrimMemory(level);
    }

    /**
     * 通过{@link #mApplicationDelegate} 获取{@link AppComponent}
     * @return 返回 {@link #mApplicationDelegate {@link #getAppComponent()}}}
     */
    @Override
    public AppComponent getAppComponent(){
        Preconditions.checkNotNull(mApplicationDelegate,"%s cannot be null",ApplicationDelegate.class.getName());
     return mApplicationDelegate.getAppComponent();
    }

    @Override
    public AndroidInjector<Object> androidInjector() {
        return mAndroidInjector;
    }

}

File: mvvmframe/src/main/kotlin/com/king/frame/mvvmframe/base/BaseApplication.kt
package com.king.frame.mvvmframe.base

import android.app.Application
import android.view.Gravity
import com.hjq.toast.Toaster

/**
 * MVVMFrame 框架基于 Google 官方的 JetPack 构建，在使用 MVVMFrame 时，需遵循一些规范：
 *
 * 你需要参照如下方式添加 @HiltAndroidApp 注解
 *
 * ```
 * // 示例
 * @HiltAndroidApp
 * class YourApplication : BaseApplication() {
 *
 * }
 * ```
 * PS：如果由于某种原因，导致你不能继承[BaseApplication]；你也可以在你自定义的Application的onCreate函
 * 数中通过调用[BaseApplication.initAppConfig]来进行初始化
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
open class BaseApplication : Application() {

    override fun onCreate() {
        super.onCreate()
        initAppConfig(this)
    }

    companion object {
        /**
         * 初始化App配置
         */
        @JvmStatic
        fun initAppConfig(application: Application) {
            // 初始化Toaster
            Toaster.init(application)
            val offsetY = application.resources.displayMetrics.heightPixels / 6
            Toaster.setGravity(Gravity.BOTTOM, 0, offsetY)
        }
    }
}
